### PR TITLE
make it possible to run tests with pytest

### PR DESCRIPTION
**Pull request by msarahan**
The main motivation here is to take advantage of pytest's excellent plugins, including pytest-xdist for parallel test running, pytest-cov for test coverage, and perhaps pytest-profiling when we want to do performance refinement. On my computer (Threadripper 3970) using pytest-xdist, I can cut the test suite from ~18 minutes down to just over 1 minute. Since multiple markdown files are running at once, some of the tests needed additional ellipsis to avoid cross-talk confusion from the global state.

Here's an initial coverage summary:

```
---------- coverage: platform linux, python 3.8.12-final-0 -----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
guild/__init__.py                                47      7    85%
guild/_api.py                                   309     11    96%
guild/_lex.py                                   690    329    52%
guild/_test.py                                  737    249    66%
guild/_yacc.py                                 1921    953    50%
guild/ansi_util.py                                6      0   100%
guild/batch_main.py                              27      0   100%
guild/batch_util.py                             355     44    88%
guild/cli.py                                    162     37    77%
guild/click_util.py                             261    105    60%
guild/cmd_impl_support.py                       148     70    53%
guild/commands/__init__.py                        0      0   100%
guild/commands/cat.py                            37      3    92%
guild/commands/cat_impl.py                       61     14    77%
guild/commands/check.py                          34      4    88%
guild/commands/check_impl.py                    376    288    23%
guild/commands/comment.py                        13      0   100%
guild/commands/compare.py                        29      0   100%
guild/commands/compare_impl.py                  393    129    67%
guild/commands/completion.py                     12      2    83%
guild/commands/completion_impl.py               120     89    26%
guild/commands/diff.py                           13      2    85%
guild/commands/diff_impl.py                     202    143    29%
guild/commands/download.py                       10      2    80%
guild/commands/download_impl.py                  24     13    46%
guild/commands/export.py                         13      2    85%
guild/commands/help.py                           20      1    95%
guild/commands/help_impl.py                      32      7    78%
guild/commands/import_.py                        13      2    85%
guild/commands/init.py                           42      8    81%
guild/commands/init_impl.py                     392    201    49%
guild/commands/install.py                        16      2    88%
guild/commands/label.py                          13      2    85%
guild/commands/ls.py                             38      2    95%
guild/commands/ls_impl.py                       146     10    93%
guild/commands/main.py                           96      0   100%
guild/commands/main_impl.py                      39      6    85%
guild/commands/mark.py                           13      2    85%
guild/commands/models.py                         13      2    85%
guild/commands/models_impl.py                    43     14    67%
guild/commands/open_.py                          32     12    62%
guild/commands/open_impl.py                      81     58    28%
guild/commands/operations.py                     19      0   100%
guild/commands/operations_impl.py                42      2    95%
guild/commands/package.py                        21      2    90%
guild/commands/package_impl.py                   39     17    56%
guild/commands/packages.py                       16      2    88%
guild/commands/packages_delete.py                16      4    75%
guild/commands/packages_impl.py                  77     52    32%
guild/commands/packages_info.py                  12      2    83%
guild/commands/packages_list.py                  11      2    82%
guild/commands/publish.py                        13      2    85%
guild/commands/publish_impl.py                   72      7    90%
guild/commands/pull.py                           13      2    85%
guild/commands/push.py                           13      2    85%
guild/commands/remote.py                         12      0   100%
guild/commands/remote_impl.py                    40     30    25%
guild/commands/remote_impl_support.py           225    169    25%
guild/commands/remote_start.py                   14      2    86%
guild/commands/remote_status.py                  13      2    85%
guild/commands/remote_stop.py                    13      2    85%
guild/commands/remote_support.py                 45     27    40%
guild/commands/remotes.py                        10      2    80%
guild/commands/remotes_impl.py                    8      3    62%
guild/commands/run.py                            93     13    86%
guild/commands/run_impl.py                     1324    295    78%
guild/commands/runs.py                           47      1    98%
guild/commands/runs_comment.py                   30     13    57%
guild/commands/runs_delete.py                    18      0   100%
guild/commands/runs_diff.py                      53      6    89%
guild/commands/runs_export.py                    18      2    89%
guild/commands/runs_impl.py                    1106    244    78%
guild/commands/runs_import.py                    18      2    89%
guild/commands/runs_info.py                      22      0   100%
guild/commands/runs_label.py                     17      2    88%
guild/commands/runs_list.py                      17      0   100%
guild/commands/runs_mark.py                      16      2    88%
guild/commands/runs_publish.py                   16      2    88%
guild/commands/runs_pull.py                      17      2    88%
guild/commands/runs_purge.py                     17      2    88%
guild/commands/runs_push.py                      21      2    90%
guild/commands/runs_restore.py                   17      2    88%
guild/commands/runs_stop.py                      17      2    88%
guild/commands/runs_support.py                  155     23    85%
guild/commands/runs_tag.py                       29     12    59%
guild/commands/s3_sync.py                        27      6    78%
guild/commands/s3_sync_impl.py                   56     36    36%
guild/commands/search.py                         11      2    82%
guild/commands/search_impl.py                    23     13    43%
guild/commands/select.py                         27      5    81%
guild/commands/server_support.py                  7      0   100%
guild/commands/service_impl_support.py           30     23    23%
guild/commands/shell.py                           9      2    78%
guild/commands/shell_impl.py                      5      1    80%
guild/commands/shutdown_timer.py                 27      6    78%
guild/commands/shutdown_timer_impl.py            80     59    26%
guild/commands/stop.py                           13      0   100%
guild/commands/sync.py                           15      2    87%
guild/commands/sync_impl.py                      76     55    28%
guild/commands/sys.py                            10      0   100%
guild/commands/tag.py                            13      2    85%
guild/commands/tensorboard.py                    26      2    92%
guild/commands/tensorboard_impl.py              115     80    30%
guild/commands/tensorflow.py                      8      0   100%
guild/commands/tensorflow_check_main.py          81     60    26%
guild/commands/tensorflow_impl.py                39     29    26%
guild/commands/tensorflow_inspect.py             13      2    85%
guild/commands/uninstall.py                      12      2    83%
guild/commands/view.py                           21      0   100%
guild/commands/view_impl.py                     243    103    58%
guild/commands/view_tester.py                    58     41    29%
guild/commands/watch.py                          17      2    88%
guild/commands/watch_impl.py                    170    131    23%
guild/config.py                                 188     15    92%
guild/entry_point_util.py                        72      6    92%
guild/exit_code.py                                6      0   100%
guild/fail.py                                     6      2    67%
guild/file_util.py                              275     27    90%
guild/flag_util.py                              259     13    95%
guild/generate_schema.py                         13      4    69%
guild/guildfile.py                             1290     84    93%
guild/help.py                                   378     47    88%
guild/index.py                                  196      7    96%
guild/init.py                                   149     79    47%
guild/ipy.py                                    473     35    93%
guild/lock.py                                    16      1    94%
guild/log.py                                     85      3    96%
guild/main.py                                    45      7    84%
guild/main_bootstrap.py                          42     18    57%
guild/model.py                                  257     44    83%
guild/model_proxy.py                             56      0   100%
guild/namespace.py                               82      8    90%
guild/op.py                                     308     32    90%
guild/op_cmd.py                                 159      4    97%
guild/op_dep.py                                 262     48    82%
guild/op_main.py                                294     49    83%
guild/op_util.py                               1134     99    91%
guild/opref.py                                   87     14    84%
guild/package.py                                 66     17    74%
guild/package_main.py                           280     83    70%
guild/pass.py                                     8      2    75%
guild/pip_util.py                               233    184    21%
guild/plugin.py                                  34      2    94%
guild/plugins/__init__.py                         0      0   100%
guild/plugins/click_flags.py                    108     14    87%
guild/plugins/config_flags.py                    95      1    99%
guild/plugins/cpu.py                             29      6    79%
guild/plugins/dask.py                            28      8    71%
guild/plugins/dask_scheduler_main.py            208    155    25%
guild/plugins/disk.py                            73     30    59%
guild/plugins/dvc.py                            208    135    35%
guild/plugins/dvc_stage_main.py                 182    136    25%
guild/plugins/dvc_util.py                       195    147    25%
guild/plugins/exec_script.py                     34      0   100%
guild/plugins/flags_import_util.py               84     11    87%
guild/plugins/gen_queue.py                      180     44    76%
guild/plugins/gpu.py                             78     46    41%
guild/plugins/hiplot.py                          58     14    76%
guild/plugins/import_argparse_flags_main.py     139     10    93%
guild/plugins/ipynb.py                          205     28    86%
guild/plugins/keras.py                           64      3    95%
guild/plugins/memory.py                          18      6    67%
guild/plugins/nbexec.py                         380    145    62%
guild/plugins/perf.py                            24     13    46%
guild/plugins/python_script.py                  295     18    94%
guild/plugins/pytorch_flags.py                   39     24    38%
guild/plugins/queue.py                           28      0   100%
guild/plugins/queue_main.py                      36      3    92%
guild/plugins/random_main.py                     36      0   100%
guild/plugins/skopt.py                           72     10    86%
guild/plugins/skopt_forest_main.py               25      0   100%
guild/plugins/skopt_gbrt_main.py                 25      0   100%
guild/plugins/skopt_gp_main.py                   20      0   100%
guild/plugins/skopt_util.py                     279     17    94%
guild/plugins/summary_util.py                   161    106    34%
guild/plugins/tensorboard.py                    126     62    51%
guild/publish.py                                473     46    90%
guild/python_util.py                            450     59    87%
guild/query/__init__.py                          65      4    94%
guild/query/qlex.py                              30      1    97%
guild/query/qparse.py                            65      0   100%
guild/remote.py                                 161     58    64%
guild/remote_run_support.py                      23      9    61%
guild/remote_util.py                            114     75    34%
guild/remotes/__init__.py                         0      0   100%
guild/remotes/azure_blob.py                     128     89    30%
guild/remotes/azure_vm.py                       216    167    23%
guild/remotes/ec2.py                            191    140    27%
guild/remotes/gist.py                           399    283    29%
guild/remotes/meta_sync.py                      139     99    29%
guild/remotes/s3.py                             181    107    41%
guild/remotes/ssh.py                            643    480    25%
guild/remotes/ssh_util.py                        72     55    24%
guild/resolver.py                               576    136    76%
guild/resource.py                                35     22    37%
guild/resourcedef.py                            219     15    93%
guild/run.py                                    190     23    88%
guild/run_check.py                               79      9    89%
guild/run_util.py                               434    167    62%
guild/run_zip_proxy.py                           62     44    29%
guild/service.py                                104     75    28%
guild/serving_util.py                            70     42    40%
guild/steps_main.py                             293     35    88%
guild/steps_util.py                              27      0   100%
guild/summary.py                                306     72    76%
guild/tabview.py                                128     82    36%
guild/tensorboard.py                            384     67    83%
guild/tfevent.py                                 85     20    76%
guild/timerange/__init__.py                       7      0   100%
guild/timerange/trlex.py                         54      0   100%
guild/timerange/trparse.py                      200      3    98%
guild/uat.py                                     86     52    40%
guild/util.py                                  1157    316    73%
guild/var.py                                    180     46    74%
guild/vcs_util.py                                85      7    92%
guild/view.py                                   257    172    33%
guild/yaml_util.py                               42      2    95%
-----------------------------------------------------------------
TOTAL                                         30768   9654    69%
```

At the time of submitting this PR, it looks like I have broken some of the tests as run by `make check`. My goal is to preserve the existing test runner as-is, but allow this as a compatible alternative test runner. With Sybil (https://sybil.readthedocs.io/en/latest/use.html) as the tool to extract examples from the doctests, it seems like we have a lot of the custom doctest capability that our built-in runner adds, such as the bash runner. The Sybil tool uses much of the existing work from the doctest customizations.

### CODE REVIEW

File: guild/plugins/python_script.py [hunk 1]
```diff
@@ -243,12 +243,13 @@ def _cache_valid(cache_path, mod_path):
...
-            and not os.getenv("_GUILD_COMPLETE")
-        ):
-            cli.note_once("Refreshing flags...")
+        # if (
+        #     os.getenv("NO_IMPORT_FLAGS_PROGRESS") != "1"
+        #     and os.getenv("FLAGS_TEST") != "1"
+        #     and not os.getenv("_GUILD_COMPLETE")
+        #     and not os.getenv("SUPPRESS_REFRESHING_FLAGS")
+        # ):
+        #     cli.note_once("Refreshing flags...")
```
Comment (msarahan) [inline guild/plugins/python_script.py]: This was showing up across many tests unexpectedly, I think because of the separate processes being used for parallel execution. In hindsight, maybe an ellipsis to catch this makes more sense.

File: guild/ipy.py [hunk 2]
```diff
@@ -438,10 +438,10 @@ def _run(op, flag_vals, opts, extra_attrs=None):
                 result = op(**flag_vals)
     except KeyboardInterrupt as e:
         exit_status = exit_code.KEYBOARD_INTERRUPT
-        util.raise_from(RunTerminated(run, e), e)
+        raise RunTerminated(run, e) from e
```
Comment (msarahan) [inline guild/ipy.py]: I've been kind of just doing things in the general direction of getting away from python 2. I will plan on extracting this kind of thing from this PR before it goes in. It should be in a dedicated PR.

File: guild/_test.py [hunk 3]
```diff
@@ -478,6 +474,8 @@ def _check_prompt_blank(lines, indent, name, lineno):
 
 
 class BashDocTestChecker(doctest.OutputChecker):
+    __test__ = False
```
Comment (msarahan) [inline guild/_test.py]: These are pytest hints, and they can go away. It's better that pytest just skip this file entirely.

File: guild/tests/ls-cmd.md [hunk 4]
```diff
@@ -104,6 +104,7 @@ List all files with `--all`.
       .guild/attrs/stopped
       .guild/attrs/user
       .guild/attrs/user_flags
+      .guild/attrs/vcs_commit
```
Comment (msarahan) [inline guild/tests/ls-cmd.md]: I don't understand why this addition came about. I'll investigate it once the rest of the errors are worked out.

File: guild/tests/run-attrs.md [hunk 5]
```diff
@@ -42,7 +42,7 @@ Non-core attributes, however, are written.
     marked: no
     label:
     sourcecode_digest: ...
-    vcs_commit:
+    vcs_commit:...
```
Comment (msarahan) [inline guild/tests/run-attrs.md]: I don't understand why this addition came about. I'll investigate it once the rest of the errors are worked out.

File: guild/tests/samples/projects/batch/guild.yml [hunk 6]
```diff
@@ -1,13 +1,5 @@
-tune:
```
Comment (msarahan) [inline guild/tests/samples/projects/batch/guild.yml]: whoa, this file went goofy. Need to revert.
